In [4]:
import numpy as np
import sympy as sp
import pickle
from IPython.display import HTML
import ipywidgets as widgets
import matplotlib as mpl
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
mpl.rcParams['legend.fontsize'] = 10
import pandas as pd
import itertools
pd.set_option('display.max_colwidth', None)

#

# function to print latex
def renderListToLatex(e):
    latex_rendering = []

    for i in range(len(e)):
        latex_rendering.append("$$" + sp.latex(e[i]) + "$$<br/>")
    
    return(HTML("".join(latex_rendering[0:])))

### Solving Polynomial Equations (15)

<hr/>

<b>Aim</b>: Explore the quartic equation and bBiTriQuad numbers

<hr/>


NOte that BiTri array actually on theOEIS at A104978 - note this is a different way of freading numbers off


Note this array allso - seen John Zhau - Fat and Thin Emergent Geometries of Hermitial One-Matrix Models (math physics) - so some connections here



Observe: implementation of $C$

Let $F1$ for a function that implements $C$.

In [2]:
def F1(m2, m3, returnCoefficientsOnly = False, returnCoefficientsOnlyWithoutSigns = True, returnCoefficientsAsFactorialStrings = False):
    c_0, c_1, c_2, c_3 = sp.symbols('c_0, c_1, c_2, c_3')
    s1 = (-1)**(m3 + 1)
    s2 = sp.factorial(2 * m2 + 3 * m3)
    s3 = sp.factorial(1 + m2 + 2 * m3) * sp.factorial(m2) * sp.factorial(m3)

    
    s4 = c_0**(1 + m2 + 2 * m3) * c_2**m2 *c_3**m3
    s5 = c_1**(2 * m2 + 3 * m3 + 1)
    
    s7 = str(2 * m2 + 3 * m3) + "!"
    s8 = str(1 + m2 + 2 * m3) + "!" + str(m2) + "!"  + str(m3) + "!"
    
    if returnCoefficientsOnly:
        s6 = s1 * (s2 / s3)
    elif returnCoefficientsOnlyWithoutSigns:
        s6 = (s2 / s3)
    elif returnCoefficientsAsFactorialStrings:
        s6 = str(s7 + " | " + s8)
    else:
        s6 = s1 * (s2 / s3) * (s4 / s5)

    return(s6)

Create variables 

In [8]:
c_0, c_1, c_2, c_3, c_4, x, t, a_0, a_1, a_2, a_3, a_4, a_5, a_6, a_7, a_8, a_9, x_1, x_2, s_1, s_2 = sp.symbols('c_0, c_1, c_2, c_3, c_4, x, t, a_0, a_1, a_2, a_3, a_4, a_5, a_6, a_7, a_8, a_9, x_1, x_2, s_1, s_2')

Let $P1$, $P2$, $P3$ be from sp 6 - general solutinos

In [9]:
with open('./SharedOutputs/SPE6.pickle', 'rb') as input:
    imported_outputs_from_another_notebook = pickle.load(input)
    
P1 = imported_outputs_from_another_notebook['xSubstitution']
P2 = imported_outputs_from_another_notebook['generalQuarticEquation']
P3 = imported_outputs_from_another_notebook['solutionForXForQuartic']


Observe: SP6 took general quartic:

In [13]:
P2

Eq(c_0 + c_1*x + c_2*x**2 + c_3*x**3 + c_4*x**4, 0)

Observe: c0 was sub for t and replaced by power series

In [10]:
P1

a_0 + a_1*t + a_2*t**2 + a_3*t**3 + a_4*t**4 + a_5*t**5 + a_6*t**6 + a_7*t**7 + a_8*t**8

Observe: a solutino for x was found

In [12]:
P3

-45*c_0**8*c_2*c_4**2/c_1**11 - 45*c_0**8*c_3**2*c_4/c_1**11 + 495*c_0**8*c_2**2*c_3*c_4/c_1**12 + 165*c_0**8*c_2*c_3**3/c_1**12 - 495*c_0**8*c_2**4*c_4/c_1**13 - 990*c_0**8*c_2**3*c_3**2/c_1**13 + 1287*c_0**8*c_2**5*c_3/c_1**14 - 429*c_0**8*c_2**7/c_1**15 - 4*c_0**7*c_4**2/c_1**9 + 72*c_0**7*c_2*c_3*c_4/c_1**10 + 12*c_0**7*c_3**3/c_1**10 - 120*c_0**7*c_2**3*c_4/c_1**11 - 180*c_0**7*c_2**2*c_3**2/c_1**11 + 330*c_0**7*c_2**4*c_3/c_1**12 - 132*c_0**7*c_2**6/c_1**13 + 7*c_0**6*c_3*c_4/c_1**8 - 28*c_0**6*c_2**2*c_4/c_1**9 - 28*c_0**6*c_2*c_3**2/c_1**9 + 84*c_0**6*c_2**3*c_3/c_1**10 - 42*c_0**6*c_2**5/c_1**11 - 6*c_0**5*c_2*c_4/c_1**7 - 3*c_0**5*c_3**2/c_1**7 + 21*c_0**5*c_2**2*c_3/c_1**8 - 14*c_0**5*c_2**4/c_1**9 - c_0**4*c_4/c_1**5 + 5*c_0**4*c_2*c_3/c_1**6 - 5*c_0**4*c_2**3/c_1**7 + c_0**3*c_3/c_1**4 - 2*c_0**3*c_2**2/c_1**5 - c_0**2*c_2/c_1**3 - c_0/c_1

Observe general form: 

$$\frac{c_0^{m_0} c_2^{m_2} c_3^{m_3} c_4^{m_4}}{c_1^{m_1}} $$

Observe 2 conditions in quartic extensino

total degree, $m_0 + m_2 + m_3 + m_4 - m_1 = 0$

index degree $ m_1 + 2m_2 + 3m_3 + 4m_4 = -1$ 

Observe - solve for 2 things: 

$$m_0 = 1 + m_2 + 2m_3 + 3m_4 $$
$$ m_1 = 1 + 2m_2 + 3m_3 + 4m_4 $$

Observe: natural extension of cubic case: 

General termm given by 


$$ f(m_0, m_3, m_4) = (-1)^{m_3 + 1} \frac{(2m_2 + 3m_3 + 4m_4)! }{(1 + 2m_2 + 3m_3 + 4m_4)!m_2!m_3!m_4!} \frac{c_0^{ 1 + m_2 + 2m_3 + 3m_4 } c_2^{m_2} c_3^{m_3} c_4^{m_4}}{c_1^{1 + 2m_2 + 3m_3 + 4m_4}} $$

Create function for this

Also change to $f(m, n, p)$ and rewrite.....

Observe - coefficients are BiTriQuad numbers.....(really taking a graph theoretic / tree theoretic) which are sub multinomial


Now define new function C(m, n, p)

Observe - looking aoriginally at BiTri array....now extending into 3 dim array, m, n, p

Natural conjecture - m, n relates to subdivision of triangles and quadrilaterals, therefore p relates to number of pentagons subvidivding polygons - so we have a polygon and this is counting number of ways to subdivide into triangles, quadrilaterals and pentagrams

jg - so heading to circle - this is possible conjecture

We can also check if they are natural numbers....

we need to expect 3d array one level at a time

Vary p - so if p = 0

make array of M(m-1, n-1, p) = so just getting BiTri nubmers

Use 8 x 8 matrix - get same array back - base level is cubic level

This suggests that all successive equatinos are nested inside each other 

look at p  = 1 (so dividing polygons into triangles, quadrilaterals, and 1 pentagon)

run p = 1

1 7 45 286  
6
28
120
etc.

Now check it in OEIS - we find that it is here, there is binomial coefficient A002694

Now grab first row

look in OEIS

Yes! - binomial A236194

check p = 2

This is nthere

Conjecture 

THe BitTriQuad = The bitriq number C(m, n, p) counts the number of a diangle subidivisonof a (2 + m + 2n + 3p) sided polygon into m triangles and n quadrialterals and p pentagons, in particular this shows that c(m, n, p) is a an integer

way of counting sides - start with 2, and then add one for every triangle we are creating, then quadrilaterals add 2 new veritces/ edges, pentagrams will add 3 vertices / edges  - if true, it is an integer in general 

Now go through all the combinations
